rclcpp/src/rclcpp/executors/static_executor_waitable.cpp

```diff
@@ -0,0 +1,264 @@
+// Copyright 2020 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/executors/static_single_threaded_executor.hpp"
+#include "rclcpp/executors/static_executor_waitable.hpp"
+
+using rclcpp::executors::StaticExecutorWaitable;
+
+StaticExecutorWaitable::~StaticExecutorWaitable()
+{
+  // Disassociate all nodes
+  for (auto & weak_node : weak_nodes_) {
+    auto node = weak_node.lock();
+    if (node) {
+      std::atomic_bool & has_executor = node->get_associated_with_executor_atomic();
+      has_executor.store(false);
+    }
+  }
+  exec_list_.clear();
+  weak_nodes_.clear();
+  guard_conditions_.clear();
+}
+
+void
+StaticExecutorWaitable::init(
+       rcl_wait_set_t* p_wait_set,
+       memory_strategy::MemoryStrategy::SharedPtr& memory_strategy,
+       StaticExecutorWaitable::SharedPtr this_shared_ptr)
+{
+  // Empty initialize executable list
+  executor::ExecutableList exec_list_();
+  // Get executor's wait_set_ pointer
+  p_wait_set_ = p_wait_set;
+  // Get executor's memory strategy ptr
+  if (memory_strategy == nullptr) {
+    throw std::runtime_error("Received NULL memory strategy in executor waitable.");
+  }
+  memory_strategy_ = memory_strategy;
+  // Get this StaticExecutorWaitable shared pointer
+  this_shared_ptr_ = this_shared_ptr;
+
+  // Get memory strategy and executable list. Prepare wait_set_
+  execute();
+}
+
+void
+StaticExecutorWaitable::execute()
+{
+  // Fill memory strategy with entities comming from weak_nodes_
+  get_memory_strategy();
+  // Fill exec_list_ with entities comming from weak_nodes_ (same as memory strategy)
+  get_executable_list();
+  // Resize the wait_set_ based on memory_strategy handles (rcl_wait_set_resize)
+  prepare_wait_set();
+}
+
+void
+StaticExecutorWaitable::get_memory_strategy()
+{
+  memory_strategy_->clear_handles();
+  bool has_invalid_weak_nodes = memory_strategy_->collect_entities(weak_nodes_);
+
+  // Clean up any invalid nodes, if they were detected
+  if (has_invalid_weak_nodes) {
+    auto node_it = weak_nodes_.begin();
+    while (node_it != weak_nodes_.end()) {
+      if (node_it->expired()) {
+        node_it = weak_nodes_.erase(node_it);
+      } else {
+        ++node_it;
+      }
+    }
+  }
+
+  // Add the static executor waitable to the memory strategy
+  memory_strategy_->add_waitable_handle(this_shared_ptr_);
```

Comment:
This can be done from the static executor itself.
during the static executor constructor/initialization, when this waitable thing is created, it is immediately added to the memory strategy.

Reply:
Yes, but then is cleared when a new entity is added: https://github.com/mauropasse/rclcpp/blob/mauro/feature/static_executor-waitable/rclcpp/src/rclcpp/executors/static_executor_waitable.cpp#L71
I could try to not delete it when rebuilding the memory_strategy..

Reply:
Is the executor that adds and remove nodes, so the operation can be done again if it's really needed

Reply:
Yes, the operation has to be done again for sure, the problem is now it's being done from the waitable and not from the executor. I should't call any executor functions from the waitable (avoid circular dependencies)

Reply:
I think your idea of not removing the waitable from the executable list should be applied here, for the memory_strategy. Also keep in mind that the order  in which entities are added to the memory strategy has to have the exact same order in the executable list. I could add the waitable first of all to both lists. Then clear all waitables but not the first

Reply:
You can't call the executor function in the waitable.
My suggestion was to do that last line from the executor itself.
However, also what you say in the second comment looks good

Reply:
I can call the last line from the executor, in the init part (I started with that approach). The problem is that all the init process has to be done again but from the `execute()` function once a new entity is added. I'll try that "not deleting" approach and let's see if it works and it's pretty.

Reply:
Okok, if it looks too complex, go ahead with the shared from this, but using the correct syntax instead of passing it from the outside

Reply:
Well now I'm using std::enable_shared_from_this, makes the code a little prettier.
Regarding the memory_strategy changes to NOT delete the waitable, was easy. But I had to add a function like:
`  void clear_handles_and_resize_waitable() override
  {
    subscription_handles_.clear();
    service_handles_.clear();
    client_handles_.clear();
    timer_handles_.clear();
    waitable_handles_.resize(1);
  }
`
which I know they won't like because it's almost duplicated from `clear_handles()`, and it's not really that necessary. What do you think?